### ISSUE TITLE
Question: Is it possible to add "fixed" request data that are always run (esp. with cli)?

### ISSUE THREAD

**Issue by hjoukl** [schemathesis/schemathesis#695]
Hi, thanks a lot for this really cool library.

(running schemathesis 2.3.3 on Python 3.6.5)

I'm having difficulties grasp if I can somehow add "fixed" data to the tests that is always run, in addition to the strategy-generated random data. I want this to ensure some valid interactions are performed against the API, e.g. actually getting a resource with a valid resource id, so that a response can get validated.

While I understand that maybe this should rather be accomplished with accompanying tests (?) of another notion (I'm new to
property-based testing) it is just too tempting to make use of schemathesis capabilities for it... ;-)

https://schemathesis.readthedocs.io/en/stable/faq.html#how-can-i-use-database-objects-ids-in-tests hints in the right direction of setting some valid ID but I'm having a hard time mapping this to cli mode.

I think a have a (kinda) working hook-based solution using along these lines:

```
# org_schemathesis.py
from typing import List
                                                                               
import schemathesis
from schemathesis.models import Case
from hypothesis import strategies
 
 
@schemathesis.hooks.register
def before_add_examples(
        context: schemathesis.hooks.HookContext,
        examples: List[Case],
    ) -> None:
    examples.append(
        Case(endpoint=context.endpoint, path_parameters={'employeeId': 12345})
        )
```

If I run this like
```
rm -fr ~/.hypothesis ./cassette.yaml && schemathesis --pre-run org_schemathesis run org-service_0.0.2.yaml --base-url=http://my.server.net:8081 --show-errors-tracebacks --auth="$ORG_AUTH" --checks=response_schema_conformance  -E /employees/ --store-network-log=cassette.yaml
```
I can see my hardcoded `employeeId` put to use in the stored cassette.yaml, along with an actual response, with a 
status: 'SUCCESS' result.

Is this a proper way to do this? IIUC the example data will always get used for the tests.

Another trouble, slightly related: If I run above with `--checks=all` my hardcoded test data is still reported and still has a a proper
response status & data, but it is mysteriously marked as a 'FAILURE' now. 
I suspect this is due to the fact that I now see other failures (for status_code_conformance checks) ; I fear there's some bug in
storing the proper status for each `http_interaction ` in `cassette.yaml`.

Best regards,
Holger

**Comment by Stranger6667** [comment]
Hi Holger! 

> thanks a lot for this really cool library.

Thank you for your kind words!

There are multiple ways to add some fixed data to the testing process and at the moment for the CLI mode the following will work:

- Adding examples to the schema directly. We support the most ways specified by the spec  - https://swagger.io/docs/specification/adding-examples/ . Note, named examples via `examples` are not supported yet.
- The second way is with hooks, exactly as you implemented. 

In the CLI mode, there is no much control available for the user in comparison with plain Python tests, so the intended way to extend the CLI mode behavior is with hooks.

For the Python case, the methods mentioned above will work but also it is possible to:

- Add `hypothesis.example()` decorator to a test function. See usage in [the docs](https://hypothesis.readthedocs.io/en/latest/reproducing.html?highlight=hypothesis.example#hypothesis.example)
- Manually update the case instance with some hard-coded values. Probably not that helpful in general, but it was the only available approach in the early days of the project and still could be applied

> I suspect this is due to the fact that I now see other failures (for status_code_conformance checks) ; I fear there's some bug in
storing the proper status for each http_interaction in cassette.yaml.

Interesting! Is there any output from the other checks? I will check this in the next following days

Cheers,
Dmitry

**Comment by hjoukl** [comment]
Dear Dmitry,
thanks for your instant response!

> There are multiple ways to add some fixed data to the testing process and at the moment for the CLI mode the following will work:
>
>    - Adding examples to the schema directly. We support the most ways specified by the spec - https://swagger.io/docs/specification/adding-examples/ . Note, named examples via examples are not supported yet.

Yeah, I gathered that from the docs but I don't usually have control over the OAS schema / wouldn't want to modify it just for this purpose.

>    - The second way is with hooks, exactly as you implemented.

Perfect! Works for me. I just had a bit of a hard time of finding this approach in the docs so I wondered if there's something more straightforward.

>>   I suspect this is due to the fact that I now see other failures (for status_code_conformance checks) ; I fear there's some bug in >>   storing the proper status for each http_interaction in cassette.yaml.
>
> Interesting! Is there any output from the other checks? I will check this in the next following days

Well some of the other checks fail with e.g. `status_code_conformance` failures, when run with `--checks=all` instead, like

```
2. Received a response with a status code, which is not defined in the schema: 404

Declared status codes: 200, 204, 400, 503

Check           : status_code_conformance
Path parameters : {'employeeId': '%C2%85'}
Headers         : {'Authorization': 'Basic Og=='}

Run this Python code to reproduce this failure: ...
```

I could try to hack up some small sample/gist e.g. with fastapi that demonstrates the behaviour but probably won't be able to do so before mid-September due to holiday reasons :-).

Thanks again, Holger

**Comment by Stranger6667** [comment]
Hi Holger,

As I see, the status conformance error is right since the 404 code is not declared in the schema. The final status is calculated from all included checks - if any check is failed, then the status is "FAILURE". It doesn't necessarily mean that the response should have 500 status code - it is only one of the available checks.

I think we can do the following:
- #702. It will make the information in cassettes more obvious - e.g., why the particular test was marked as "FAILURE";
- Document this behavior in our docs;

Let me know if we can do something more to improve this area :)

**Comment by hjoukl** [comment]
Hi Dmitry,

by final status, do you mean that if a single request GET /endpoint/<param> is performed, than this will get checked for

- not_a_server_error
- status_code_conformance
- content_type_conformance
- response_schema_conformance

and if any these checks for one single endpoint invocation fail then it is status: 'FAILURE'?

Makes sense. Yes, I suppose #702 would then provide helpful insight into the failure reason.

But I think I have a slightly different problem: 

I force-add a fixed example to the test samples by pre-run hook, e.g. to make sure certain known values get "triggered". I run with `--checks=response_schema_conformance` i.e. exclude `status_code_conformance` checking (amongst others). 
I run this against a service with status conformance problems - all tests are successful, as these errors aren't checked for.
I then rerun all checks and expect some tests to fail, but I'd also expect some tests to be successful.

Minimal example using fastapi:

A simple fastapi hello-world-service. The marked lines have been added after an initial retrieval of its openapi.json:
```
# main.py
from fastapi import FastAPI, HTTPException
from pydantic import BaseModel


app = FastAPI(title='hello-world-service')


# data model
class HelloResponse(BaseModel):
    message: str


@app.get("/hello", response_model=HelloResponse)
def read_hello():
    return {"message": "Hello, world!"}


@app.get("/hello/{name}", response_model=HelloResponse)
def read_hello_name(name: str):
    ###### 
    # These lines have been added after retrieving openapi.json from fastapi service,
    # so they provoke a status result that is definitely not documented in our oas spec.
    if name == '0':
        raise HTTPException(status_code=404, detail="not found")
    ###### 
    return {"message": f"Hello, {name}!"}
```

My hook:
```
# add_examples.py
from typing import List

import schemathesis
from schemathesis.models import Case


@schemathesis.hooks.register
def before_add_examples(
        context: schemathesis.hooks.HookContext,
        examples: List[Case],
    ) -> None:
    examples.append(
        Case(endpoint=context.endpoint, path_parameters={'name': 'holger'})
        )

```

Run test subset:
```
rm -fr ~/.hypothesis ./cassette.yaml && ~/pydev/venv/schemathesis-venv/bin/schemathesis --pre-run add_examples run openapi.json --base-url=http://myserver.net:8000 --show-errors-tracebacks --checks=response_schema_conformance  -E /hello/ --store-network-log=cassette.yaml
======================= Schemathesis test session starts ======================
platform Linux -- Python 3.6.5, schemathesis-2.3.3, hypothesis-5.23.3, hypothesis_jsonschema-0.17.3, jsonschema-3.2.0
rootdir: ...
Specification version: Open API 3.0.2
Workers: 1
collected endpoints: 1

GET /hello/{name} .                                                     [100%]

=================================== SUMMARY ===================================

Performed checks:
    response_schema_conformance                    101 / 101 passed          PASSED 

Network log: cassette.yaml
```

All tests are reported as succesful, including my custom added example:
```
$ grep "status: 'SUCCESS'" cassette.yaml |wc -l
101
```

Now rerun with --checks=all:

```
$ rm -fr ~/.hypothesis ./cassette.yaml && ~/pydev/venv/schemathesis-venv/bin/schemathesis --pre-run add_examples run openapi.json --base-url=http://myserver.net:8000 --show-errors-tracebacks --checks=all  -E /hello/ --store-network-log=cassette.yaml
======================= Schemathesis test session starts ======================
platform Linux -- Python 3.6.5, schemathesis-2.3.3, hypothesis-5.23.3, hypothesis_jsonschema-0.17.3, jsonschema-3.2.0
rootdir: ...
Specification version: Open API 3.0.2
Workers: 1
collected endpoints: 1

GET /hello/{name} F                                                     [100%]

=================================== FAILURES ==================================
______________________________ GET: /hello/{name} _____________________________
1. Received a response with a status code, which is not defined in the schema: 404

Declared status codes: 200, 422

Check           : status_code_conformance
Path parameters : {'name': '0'}

Run this Python code to reproduce this failure: 

    requests.get('http://myserver.net:8000/hello/0')

Or add this option to your command line parameters: --hypothesis-seed=87345815337462284460351829168824299911
=================================== SUMMARY ===================================

Performed checks:
    not_a_server_error                             4 / 4 passed          PASSED 
    status_code_conformance                        1 / 4 passed          FAILED 
    content_type_conformance                       4 / 4 passed          PASSED 
    response_schema_conformance                    4 / 4 passed          PASSED 

Network log: cassette.yaml
```

Now, all tests are reported as failures; this includes my custom fixed example:

```
command: 'schemathesis --pre-run add_examples run openapi.json --base-url=http://myserver.net:8000 --show-errors-tracebacks --checks=all -E /hello/ --store-network-log=cassette.yaml'
recorded_with: 'Schemathesis 2.3.3'
http_interactions:
- id: '1'
  status: 'FAILURE'
  seed: '138515993875395522801610595088206829211'
  elapsed: '0.006496'
  recorded_at: '2020-10-02T18:41:46.480052'
  request:
    uri: 'http://myserver.net:8000/hello/holger'
    method: 'GET'
    headers:
      User-Agent:
      - 'schemathesis/2.3.3'
      Accept-Encoding:
      - 'gzip, deflate'
      Accept:
      - '*/*'
      Connection:
      - 'keep-alive'
    body:
      encoding: 'utf-8'
      base64_string: ''
  response:
    status:
      code: '200'
      message: 'OK'
    headers:
      date:
      - 'Fri, 02 Oct 2020 16:41:45 GMT'
      server:
      - 'uvicorn'
      content-length:
      - '28'
      content-type:
      - 'application/json'
    body:
      encoding: 'utf8'
      base64_string: 'eyJtZXNzYWdlIjoiSGVsbG8sIGhvbGdlciEifQ=='
    http_version: '1.1'
...
```

I'd say this clearly should still be 'SUCCESS'. Of course, some of the other tests are bound to fail :-).

Hm, on a sidenote: Why does schemathesis run only 4 tests in the latter case? Something to do with hypothesis/property based testing works generally?

Best regards,
Holger

**Comment by Stranger6667** [comment]
Hi, Holger!

Thank you for providing such a detailed report! :+1:  I will take a look at it tomorrow and respond as soon as I have more info!

> and if any these checks for one single endpoint invocation fail then it is status: 'FAILURE'?

Yes, that's correct

> Why does schemathesis run only 4 tests in the latter case? Something to do with hypothesis/property based testing works generally?

It is the default Hypothesis behavior in such cases - if an error is found, then it often stops searching for new cases (especially if the found error is already close to its possible minimal form)

**Comment by hjoukl** [comment]
Great! Not urgent at all for me, btw.

> Thank you for providing such a detailed report

A pleasure! 😀

One thing I wanted to mention but forgot: Note how the stdout/stderr output reports successful tests but in the *cassette* everything's marked as a failure.

And thanks for the hypothesis fail-fast behaviour explanation, much appreciated.

**Comment by Stranger6667** [comment]
You are very welcome!

> I will take a look at it tomorrow and respond as soon as I have more info!

It turned out that this "status" value comes from the overall test item outcome, not an individual response validation, which is misleading and wasn't intended to work like this. Will fix it soon :)

**Comment by Stranger6667** [comment]
#702 is resolved and #779 is merged :) I will release these changes next week :)

**Comment by Stranger6667** [comment]
Thank you for your patience and time, Holger :)
